File: cuhara.qua.back/src/core/cqrs/command.ts
export interface Command<TResponse = void> {
  readonly type: string;
}

export interface CommandHandler<
  TCommand extends Command<TResponse>,
  TResponse = void
> {
  handle(command: TCommand): Promise<TResponse>;
}

export abstract class BaseCommand<TResponse = void> implements Command<TResponse> {
  abstract readonly type: string;
}

File: cuhara.qua.go/internal/common/cqrs/command.go
package cqrs

import "context"

type Command[T any] interface {
	Type() string
}

type CommandHandler[T any] interface {
	Handle(ctx context.Context, command Command[T]) (T, error)
}

type BaseCommand[T any] struct {
	commandType string
}

func NewBaseCommand[T any](commandType string) BaseCommand[T] {
	return BaseCommand[T]{
		commandType: commandType,
	}
}

func (c BaseCommand[T]) Type() string {
	return c.commandType
}
